File: backend/app/routers/roadmap.py
"""
Study Roadmap API routes.

Provides endpoints for:
- Creating personalized study roadmaps
- Getting user's active roadmap with milestones
- Getting daily study plans
- Updating milestone progress
- Adapting roadmap based on performance
"""

from datetime import datetime
from typing import Any

from fastapi import APIRouter, Depends, Header, HTTPException
from pydantic import BaseModel, Field
from sqlalchemy import select
from sqlalchemy.orm import Session

from app.database import get_db
from app.models.roadmap import MilestoneStatus, RoadmapStatus, StudyRoadmap
from app.models.user import User
from app.services.roadmap_generator import (
    RoadmapGenerator,
    get_roadmap_with_milestones,
    get_user_roadmap,
)

router = APIRouter(prefix="/api/roadmap", tags=["roadmap"])


# ========================
# Pydantic Schemas
# ========================


class CreateRoadmapRequest(BaseModel):
    """Request schema for creating a new roadmap."""

    exam_date: datetime = Field(..., description="Target exam date")
    weekly_study_hours: int = Field(
        default=10, ge=1, le=40, description="Hours available for study per week"
    )
    study_days_per_week: int = Field(
        default=5, ge=1, le=7, description="Days available for study per week"
    )


class DailyPlanActivity(BaseModel):
    """An activity in the daily plan."""

    type: str
    count: int | None = None
    duration_minutes: int | None = None
    full: bool | None = None


class DailyPlan(BaseModel):
    """Daily study plan."""

    type: str
    hours: float | None = None
    domain: str | None = None
    activities: list[DailyPlanActivity] | None = None


class MilestoneResponse(BaseModel):
    """Response schema for a milestone."""

    id: int
    title: str
    description: str | None
    week_number: int
    scheduled_date: str
    target_date: str
    domain_ids: list[int]
    daily_plan: dict[str, DailyPlan]
    completion_criteria: dict[str, Any]
    status: str
    flashcards_completed: int
    questions_completed: int
    quiz_score: float | None
    completed_at: str | None


class RoadmapResponse(BaseModel):
    """Response schema for a study roadmap."""

    id: int
    user_id: str
    exam_date: str
    weekly_study_hours: int
    study_days_per_week: int
    status: str
    focus_areas: list[int] | None
    recommendations: dict[str, Any] | None
    total_milestones: int
    completed_milestones: int
    created_at: str
    updated_at: str
    last_adapted_at: str | None
    milestones: list[MilestoneResponse]


class UpdateMilestoneRequest(BaseModel):
    """Request schema for updating milestone progress."""

    flashcards_completed: int | None = None
    questions_completed: int | None = None
    mark_complete: bool = False


class DailyPlanResponse(BaseModel):
    """Response schema for daily plan."""

    date: str
    plan: DailyPlan | None


class AdaptRoadmapResponse(BaseModel):
    """Response schema for roadmap adaptation."""

    message: str
    roadmap_id: int
    adapted: bool


# ========================
# Helper Functions
# ========================


async def get_or_create_user(db: Session, anonymous_id: str) -> User:
    """Get existing user or create new one based on anonymous_id."""
    user = db.execute(
        select(User).where(User.anonymous_id == anonymous_id)
    ).scalar_one_or_none()

    if not user:
        user = User(anonymous_id=anonymous_id)
        db.add(user)
        db.commit()
        db.refresh(user)

    return user


# ========================
# API Endpoints
# ========================


@router.post("/create", response_model=RoadmapResponse)
async def create_roadmap(
    request: CreateRoadmapRequest,
    x_anonymous_id: str = Header(..., alias="X-Anonymous-ID"),
    db: Session = Depends(get_db),
) -> RoadmapResponse:
    """
    Create a new personalized study roadmap.

    Generates a complete study plan with weekly milestones based on:
    - Target exam date
    - Available study time
    - User's current weak domains

    The roadmap adapts the curriculum across weeks following PMP 2026 ECO
    domain weights.
    """
    user = await get_or_create_user(db, x_anonymous_id)

    generator = RoadmapGenerator(db)
    roadmap = generator.create_roadmap(
        user_id=user.id,
        exam_date=request.exam_date,
        weekly_study_hours=request.weekly_study_hours,
        study_days_per_week=request.study_days_per_week,
    )

    roadmap_data = get_roadmap_with_milestones(db, roadmap.id, user.id)
    return RoadmapResponse(**roadmap_data)


@router.get("/active", response_model=RoadmapResponse)
async def get_active_roadmap(
    x_anonymous_id: str = Header(..., alias="X-Anonymous-ID"),
    db: Session = Depends(get_db),
) -> RoadmapResponse:
    """
    Get the user's active study roadmap.

    Returns the complete roadmap with all milestones including:
    - Weekly study plans
    - Daily breakdown
    - Progress tracking
    - AI-generated recommendations
    """
    user = await get_or_create_user(db, x_anonymous_id)

    roadmap = get_user_roadmap(db, user.id)

    if not roadmap:
        raise HTTPException(
            status_code=404,
            detail="No active roadmap found. Create one first.",
        )

    roadmap_data = get_roadmap_with_milestones(db, roadmap.id, user.id)
    return RoadmapResponse(**roadmap_data)


@router.get("/{roadmap_id}", response_model=RoadmapResponse)
async def get_roadmap(
    roadmap_id: int,
    x_anonymous_id: str = Header(..., alias="X-Anonymous-ID"),
    db: Session = Depends(get_db),
) -> RoadmapResponse:
    """
    Get a specific roadmap by ID.

    Returns the complete roadmap with all milestones.
    """
    user = await get_or_create_user(db, x_anonymous_id)

    roadmap = db.execute(
        select(StudyRoadmap)
        .where(StudyRoadmap.id == roadmap_id)
        .where(StudyRoadmap.user_id == str(user.id))
    ).scalar_one_or_none()

    if not roadmap:
        raise HTTPException(status_code=404, detail="Roadmap not found")

    roadmap_data = get_roadmap_with_milestones(db, roadmap_id, user.id)
    return RoadmapResponse(**roadmap_data)


@router.get("/{roadmap_id}/daily/{date}", response_model=DailyPlanResponse)
async def get_daily_plan_endpoint(
    roadmap_id: int,
    date: str,
    x_anonymous_id: str = Header(..., alias="X-Anonymous-ID"),
    db: Session = Depends(get_db),
) -> DailyPlanResponse:
    """
    Get the daily study plan for a specific date.

    Returns the planned activities for the given date including:
    - Study hours
    - Focus domains
    - Flashcard and question targets
    """
    user = await get_or_create_user(db, x_anonymous_id)

    # Verify roadmap ownership
    roadmap = db.execute(
        select(StudyRoadmap)
        .where(StudyRoadmap.id == roadmap_id)
        .where(StudyRoadmap.user_id == str(user.id))
    ).scalar_one_or_none()

    if not roadmap:
        raise HTTPException(status_code=404, detail="Roadmap not found")

    try:
        target_date = datetime.fromisoformat(date.replace("Z", "+00:00"))
    except ValueError:
        raise HTTPException(status_code=400, detail="Invalid date format")

    generator = RoadmapGenerator(db)
    plan = generator.get_daily_plan(roadmap_id, target_date)

    return DailyPlanResponse(
        date=date,
        plan=DailyPlan(**plan) if plan else None,
    )


@router.put("/{roadmap_id}/milestones/{milestone_id}")
async def update_milestone(
    roadmap_id: int,
    milestone_id: int,
    request: UpdateMilestoneRequest,
    x_anonymous_id: str = Header(..., alias="X-Anonymous-ID"),
    db: Session = Depends(get_db),
) -> MilestoneResponse:
    """
    Update progress for a specific milestone.

    Use this to track completed flashcards and questions for a milestone.
    Can also mark the milestone as complete.
    """
    user = await get_or_create_user(db, x_anonymous_id)

    # Verify roadmap ownership
    roadmap = db.execute(
        select(StudyRoadmap)
        .where(StudyRoadmap.id == roadmap_id)
        .where(StudyRoadmap.user_id == str(user.id))
    ).scalar_one_or_none()

    if not roadmap:
        raise HTTPException(status_code=404, detail="Roadmap not found")

    generator = RoadmapGenerator(db)
    milestone = generator.update_milestone_progress(
        milestone_id=milestone_id,
        user_id=user.id,
        flashcards_completed=request.flashcards_completed,
        questions_completed=request.questions_completed,
        mark_complete=request.mark_complete,
    )

    return MilestoneResponse(
        id=milestone.id,
        title=milestone.title,
        description=milestone.description,
        week_number=milestone.week_number,
        scheduled_date=milestone.scheduled_date.isoformat(),
        target_date=milestone.target_date.isoformat(),
        domain_ids=eval(milestone.domain_ids) if milestone.domain_ids else [],
        daily_plan=eval(milestone.daily_plan) if milestone.daily_plan else {},
        completion_criteria=eval(milestone.completion_criteria) if milestone.completion_criteria else {},
        status=milestone.status,
        flashcards_completed=milestone.flashcards_completed,
        questions_completed=milestone.questions_completed,
        quiz_score=milestone.quiz_score,
        completed_at=milestone.completed_at.isoformat() if milestone.completed_at else None,
    )


@router.post("/{roadmap_id}/adapt", response_model=AdaptRoadmapResponse)
async def adapt_roadmap(
    roadmap_id: int,
    x_anonymous_id: str = Header(..., alias="X-Anonymous-ID"),
    db: Session = Depends(get_db),
) -> AdaptRoadmapResponse:
    """
    Adapt the roadmap based on current progress.

    Re-analyzes user performance and adjusts remaining milestones:
    - Allocates more time to weak domains
    - Adjusts daily plans based on completion rates
    - Updates AI recommendations

    Call this after completing major milestones or when
    performance has changed significantly.
    """
    user = await get_or_create_user(db, x_anonymous_id)

    # Verify roadmap ownership
    roadmap = db.execute(
        select(StudyRoadmap)
        .where(StudyRoadmap.id == roadmap_id)
        .where(StudyRoadmap.user_id == str(user.id))
    ).scalar_one_or_none()

    if not roadmap:
        raise HTTPException(status_code=404, detail="Roadmap not found")

    generator = RoadmapGenerator(db)
    adapted_roadmap = generator.adapt_roadmap(roadmap_id, user.id)

    return AdaptRoadmapResponse(
        message="Roadmap adapted successfully based on current progress",
        roadmap_id=adapted_roadmap.id,
        adapted=True,
    )


@router.delete("/{roadmap_id}")
async def delete_roadmap(
    roadmap_id: int,
    x_anonymous_id: str = Header(..., alias="X-Anonymous-ID"),
    db: Session = Depends(get_db),
) -> dict[str, str]:
    """
    Delete or archive a roadmap.

    Soft-deletes by archiving the roadmap rather than removing it.
    """
    user = await get_or_create_user(db, x_anonymous_id)

    roadmap = db.execute(
        select(StudyRoadmap)
        .where(StudyRoadmap.id == roadmap_id)
        .where(StudyRoadmap.user_id == str(user.id))
    ).scalar_one_or_none()

    if not roadmap:
        raise HTTPException(status_code=404, detail="Roadmap not found")

    # Archive instead of delete
    roadmap.status = RoadmapStatus.ARCHIVED
    db.commit()

    return {"message": "Roadmap archived successfully"}

File: frontend/src/lib/api/roadmap.ts
/**
 * Roadmap API client
 *
 * Provides functions for interacting with the study roadmap endpoints:
 * - POST /api/roadmap/create - Create a new study roadmap
 * - GET /api/roadmap/active - Get the user's active roadmap
 * - GET /api/roadmap/:id - Get a specific roadmap
 * - GET /api/roadmap/:id/daily/:date - Get daily study plan
 * - PUT /api/roadmap/:id/milestones/:milestoneId - Update milestone progress
 * - POST /api/roadmap/:id/adapt - Adapt roadmap based on progress
 * - DELETE /api/roadmap/:id - Archive a roadmap
 */

import { storage } from './client';

// ============ Types ============

/** Roadmap status */
export type RoadmapStatus = 'active' | 'paused' | 'completed' | 'archived';

/** Milestone status */
export type MilestoneStatus = 'pending' | 'in_progress' | 'completed' | 'skipped';

/** Activity in a daily study plan */
export interface DailyPlanActivity {
  type: string;
  count?: number;
  duration_minutes?: number;
  full?: boolean;
}

/** Daily study plan for a specific day */
export interface DailyPlan {
  type: string;
  hours?: number;
  domain?: string;
  activities?: DailyPlanActivity[];
}

/** A milestone in the study roadmap */
export interface RoadmapMilestone {
  id: number;
  title: string;
  description: string | null;
  week_number: number;
  scheduled_date: string;
  target_date: string;
  domain_ids: number[];
  daily_plan: Record<string, DailyPlan>;
  completion_criteria: Record<string, unknown>;
  status: MilestoneStatus;
  flashcards_completed: number;
  questions_completed: number;
  quiz_score: number | null;
  completed_at: string | null;
}

/** Complete study roadmap */
export interface StudyRoadmap {
  id: number;
  user_id: string;
  exam_date: string;
  weekly_study_hours: number;
  study_days_per_week: number;
  status: RoadmapStatus;
  focus_areas: number[] | null;
  recommendations: Record<string, unknown> | null;
  total_milestones: number;
  completed_milestones: number;
  created_at: string;
  updated_at: string;
  last_adapted_at: string | null;
  milestones: RoadmapMilestone[];
}

/** Request to create a new roadmap */
export interface CreateRoadmapRequest {
  exam_date: string;
  weekly_study_hours: number;
  study_days_per_week: number;
}

/** Request to update milestone progress */
export interface UpdateMilestoneRequest {
  flashcards_completed?: number;
  questions_completed?: number;
  mark_complete?: boolean;
}

/** Response for daily plan endpoint */
export interface DailyPlanResponse {
  date: string;
  plan: DailyPlan | null;
}

/** Response for adapt roadmap endpoint */
export interface AdaptRoadmapResponse {
  message: string;
  roadmap_id: number;
  adapted: boolean;
}

const ROADMAP_KEY = 'pmp_study_roadmap';

// ============ API Functions (Mocked) ============

export async function createRoadmap(request: CreateRoadmapRequest): Promise<StudyRoadmap> {
  const now = new Date().toISOString();
  const roadmap: StudyRoadmap = {
    id: 1,
    user_id: 'guest',
    exam_date: request.exam_date,
    weekly_study_hours: request.weekly_study_hours,
    study_days_per_week: request.study_days_per_week,
    status: 'active',
    focus_areas: [],
    recommendations: {},
    total_milestones: 4,
    completed_milestones: 0,
    created_at: now,
    updated_at: now,
    last_adapted_at: null,
    milestones: [
      {
        id: 1,
        title: 'Foundations & People Domain',
        description: 'Focus on leadership and team dynamics',
        week_number: 1,
        scheduled_date: now,
        target_date: now,
        domain_ids: [1],
        daily_plan: {},
        completion_criteria: {},
        status: 'in_progress',
        flashcards_completed: 0,
        questions_completed: 0,
        quiz_score: null,
        completed_at: null,
      }
    ],
  };
  storage.set(ROADMAP_KEY, roadmap);
  return roadmap;
}

export async function getActiveRoadmap(): Promise<StudyRoadmap> {
  const roadmap = storage.get<StudyRoadmap>(ROADMAP_KEY);
  if (!roadmap) throw new Error('No active roadmap');
  return roadmap;
}

export async function getRoadmap(_roadmapId: number): Promise<StudyRoadmap> {
  if (!_roadmapId) return {} as StudyRoadmap;
  return getActiveRoadmap();
}

export async function getDailyPlan(_roadmapId: number, date: string): Promise<DailyPlanResponse> {
  if (!_roadmapId) return { date, plan: null };
  return {
    date,
    plan: {
      type: 'study',
      hours: 2,
      domain: 'People',
      activities: [
        { type: 'flashcards', count: 20 },
        { type: 'questions', count: 10 }
      ]
    }
  };
}

export async function updateMilestone(
  _roadmapId: number,
  milestoneId: number,
  request: UpdateMilestoneRequest
): Promise<RoadmapMilestone> {
  if (!_roadmapId) return {} as RoadmapMilestone;
  const roadmap = await getActiveRoadmap();
  const milestone = roadmap.milestones.find((m: RoadmapMilestone) => m.id === milestoneId);
  if (milestone) {
    if (request.flashcards_completed !== undefined) milestone.flashcards_completed = request.flashcards_completed;
    if (request.questions_completed !== undefined) milestone.questions_completed = request.questions_completed;
    if (request.mark_complete) milestone.status = 'completed';
    storage.set(ROADMAP_KEY, roadmap);
  }
  return milestone!;
}

export async function adaptRoadmap(roadmapId: number): Promise<AdaptRoadmapResponse> {
  return {
    message: 'Roadmap adapted locally',
    roadmap_id: roadmapId,
    adapted: true,
  };
}

export async function archiveRoadmap(_roadmapId: number): Promise<{ message: string }> {
  if (!_roadmapId) return { message: 'No roadmap specified' };
  storage.set(ROADMAP_KEY, null);
  return { message: 'Roadmap archived' };
}
